File: src/keystore2john.c
/*
 * Output password protected Java KeyStore files in JtR
 * compatible format.
 *
 * Output Format: $keystore$target$data_length$data$hash$nkeys$keylength$keydata$keylength$keydata...
 *
 * Where,
 *
 * target == 0 if container password is to be cracked
 * target == 1 if private key password(s) are to be cracked
 *
 * TODO:
 *
 * 1. Private Keys can be encrypted with a password different from the container password.
 * Add support for cracking such keys.
 *
 * 2. Add ability to select any key for cracking in case multiple keys are present.
 */

/*
 * KEYSTORE FORMAT:
 *
 * Magic number (big-endian integer),
 * Version of this file format (big-endian integer),
 *
 * Count (big-endian integer),
 * followed by "count" instances of either:
 *
 *     {
 *      tag=1 (big-endian integer),
 *      alias (UTF string)
 *      timestamp
 *      encrypted private-key info according to PKCS #8
 *          (integer length followed by encoding)
 *      cert chain (integer count, then certs; for each cert,
 *          integer length followed by encoding)
 *     }
 *
 * or:
 *
 *     {
 *      tag=2 (big-endian integer)
 *      alias (UTF string)
 *      timestamp
 *      cert (integer length followed by encoding)
 *     }
 *
 * ended by a keyed SHA1 hash (bytes only) of
 *     { password + whitener + preceding body }
 */

#include <stdarg.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>

#include "stdint.h"
#include "jumbo.h"
#include "memdbg.h"

#define N 819200

static int MAGIC = 0xfeedfeed;
static int VERSION_1 = 0x01;
static int VERSION_2 = 0x02;

static unsigned char data[N];
static unsigned char protectedPrivKey[N];
static unsigned char certdata[N];
static unsigned char buf[N];

static void warn_exit(const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	if (fmt != NULL)
		vfprintf(stderr, fmt, ap);
	va_end(ap);
	fprintf(stderr, "\n");

	exit(EXIT_FAILURE);
}

static uint32_t fget32(FILE * fp)
{
        uint32_t v = fgetc(fp) << 24;
        v |= fgetc(fp) << 16;
        v |= fgetc(fp) << 8;
        v |= fgetc(fp);
        return v;
}

static void print_hex(unsigned char *str, int len)
{
	int i;
	for (i = 0; i < len; ++i)
		printf("%02x", str[i]);
}

static void process_file(char *filename)
{
	FILE *fp;
	int i, j;
	int tag;
	unsigned char p, length;
	int count, keysize = 0;
	int numOfCerts, certsize;
	long size;
	int pos;
	unsigned char md[20];
	int xMagic;
	int xVersion;
	char *bname;
	const char *extension[] = { ".keystore" };

	if (!(fp = fopen(filename, "rb"))) {
		fprintf(stderr, "! %s: %s\n", filename, strerror(errno));
		goto bail;
	}
	fseek(fp, 0, SEEK_END);
	size = ftell(fp);
	if (size > sizeof(data))
		warn_exit("size=%ld should smaller or equal to sizeof(data)=%d",
			size, sizeof(data));
	fseek(fp, 0, SEEK_SET);
	if (fread(data, size, 1,  fp) != 1)
		warn_exit("Error: read failed.");

	fseek(fp, 0, SEEK_SET);

	xMagic = fget32(fp);
	xVersion = fget32(fp);

	if (xMagic!=MAGIC || (xVersion!=VERSION_1 && xVersion!=VERSION_2)) {
		fprintf(stderr, "Invalid keystore format\n");
		goto bail;
	}

	count = fget32(fp);
	for (i = 0; i < count; i++) {
			tag = fget32(fp);

		if (tag == 1) { // key entry
			// Read the alias
			p = fgetc(fp);
			length = fgetc(fp);

			if (sizeof(buf) < length || fread(buf, length, 1, fp) != 1)
				warn_exit("Error: read failed.");

			// Read the (entry creation) date
			if (fread(buf, 8, 1, fp) != 1)
				warn_exit("Error: read failed.");

			// Read the key
			keysize = fget32(fp);
			if (sizeof(protectedPrivKey) < keysize || fread(protectedPrivKey, keysize, 1, fp) != 1)
				warn_exit("Error: read failed.");

			// read certificates
			numOfCerts = fget32(fp);
			if (numOfCerts > 0) {
				for (j = 0; j < numOfCerts; j++) {
					if (xVersion == 2) {
						// read the certificate type
						p = fgetc(fp);
						if (p != 1 && p != 0)
							warn_exit("Error: p=%d which should be 1 or 0.", p);
						length = fgetc(fp);
						if (sizeof(buf) < length || fread(buf, length, 1, fp) != 1)
							warn_exit("Error: read failed.");
					}
					// read certificate data
					certsize = fget32(fp);
					if (sizeof(certdata) < certsize || fread(certdata, certsize, 1, fp) != 1)
						warn_exit("Error: read failed.");
				}
			}
			// We can be sure now that numOfCerts of certs are read
		} else if (tag == 2) { // trusted certificate entry
			// Read the alias
			p = fgetc(fp);
			length = fgetc(fp);
			if (sizeof(buf) < length || fread(buf, length, 1, fp) != 1)
				warn_exit("Error: read failed.");

			// Read the (entry creation) date
			if (fread(buf, 8, 1, fp) != 1)
				warn_exit("Error: read failed.");

			// Read the trusted certificate
			if (xVersion == 2) {
				// read the certificate type
				p = fgetc(fp);
				length = fgetc(fp);
				if (sizeof(buf) < length || fread(buf, length, 1, fp) != 1)
					warn_exit("Error: read failed.");
			}
			certsize = fget32(fp);
			if (sizeof(certdata) < certsize || fread(certdata, certsize, 1, fp) != 1)
				warn_exit("Error: read failed.");
		} else {
			fprintf(stderr, "Unrecognized keystore entry");
			fclose(fp);
			goto bail;
		}
	}

	/* how much data have we processed */
	pos = ftell(fp);

	/* read hash */
	if (fread(md, 20, 1, fp) != 1)
		warn_exit("Error: read failed.");

	bname = strip_suffixes(basename(filename), extension, 1);
	printf("%s:$keystore$0$%d$", bname, pos);
	print_hex(data, pos);
	printf("$");
	print_hex(md, 20);
	printf("$%d$%d$", count, keysize);
	print_hex(protectedPrivKey, keysize);
	printf(":::::%s\n", filename);
bail:
	return;
}

int keystore2john(int argc, char **argv)
{
        if (argc < 2) {
                fprintf(stderr, "Usage: %s <.keystore file>\n", argv[0]);
                exit(-1);
        }
        process_file(argv[1]);
	MEMDBG_PROGRAM_EXIT_CHECKS(stderr);
	return 0;
}

File: run/keystore2john.py
#!/usr/bin/env python

# This software is Copyright (c) 2014, Sanju Kholia <sanju.kholia at gmail.com>
# and it is hereby released to the general public under the following terms:
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted.
#
# Currently, only Sun "JKS" keystore files are supported.
# See https://www.bouncycastle.org/specifications.html for details.
#
# Use "Portecle" (https://sourceforge.net/projects/portecle/files/) for testing
# various keystore file formats.

"""

Output password protected Java KeyStore files in JtR compatible format.

Output Format: $keystore$target$data_length$data$hash$nkeys$keylength$keydata$keylength$keydata...

Where,

target == 0 if container password is to be cracked
target == 1 if private key password(s) are to be cracked

KEYSTORE FORMAT:

Magic number (big-endian integer),
Version of this file format (big-endian integer),

Count (big-endian integer),
followed by "count" instances of either:

    {
     tag=1 (big-endian integer),
     alias (UTF string)
     timestamp
     encrypted private-key info according to PKCS #8
         (integer length followed by encoding)
     cert chain (integer count, then certs; for each cert,
         integer length followed by encoding)
    }

or:

    {
     tag=2 (big-endian integer)
     alias (UTF string)
     timestamp
     cert (integer length followed by encoding)
    }

ended by a keyed SHA1 hash (bytes only) of
    { password + whitener + preceding body }

See http://metastatic.org/source/JKS.java (implementation of the "JKS" key
store) for more details.

"""

import sys
import os
import struct
from binascii import hexlify

MAGIC = 0xfeedfeed
VERSION_1 = 0x01
VERSION_2 = 0x02


def read_utf_string(fd):
    buf = fd.read(2)
    if len(buf) < 2:
        return b""
    length = struct.unpack(">H", buf)[0]
    return fd.read(length)


def process_file(filename):
    try:
        fd = open(filename, "rb")
    except IOError:
        e = sys.exc_info()[1]
        sys.stderr.write("! %s: %s\n" % (filename, str(e)))
        return

    # read the entire file into data variable
    data = fd.read()
    fd.seek(0, os.SEEK_SET)

    # start actual processing
    buf = fd.read(4)
    if len(buf) < 4:
        return
    xMagic = struct.unpack("> I", buf)[0]
    buf = fd.read(4)
    if len(buf) < 4:
        return
    xVersion = struct.unpack("> I", buf)[0]

    if (xMagic != MAGIC or (xVersion != VERSION_1 and xVersion != VERSION_2)):
        sys.stderr.write("%s: Invalid keystore format\n" % filename)
        return

    buf = fd.read(4)
    if len(buf) < 4:
        return
    count = struct.unpack("> I", buf)[0]

    keys = []
    for i in range(0, count):
        buf = fd.read(4)
        if len(buf) < 4:
            break
        tag = struct.unpack("> I", buf)[0]

        if (tag == 1):  # key entry
            # Read the alias
            alias = read_utf_string(fd)

            # Read the (entry creation) date
            buf = fd.read(8)
            assert(len(buf) == 8)

            # Read the key
            buf = fd.read(4)
            keysize = struct.unpack("> I", buf)[0]
            protectedPrivKey = fd.read(keysize)
            assert(len(protectedPrivKey) == keysize)

            # read certificates
            buf = fd.read(4)
            numOfCerts = struct.unpack("> I", buf)[0]
            if (numOfCerts > 0):
                for j in range(0, numOfCerts):
                    if xVersion == 2:
                        # read the certificate type
                        read_utf_string(fd)

                    # read certificate data
                    buf = fd.read(4)
                    certsize = struct.unpack("> I", buf)[0]
                    certdata = fd.read(certsize)
                    assert(len(certdata) == certsize)

            keys.append((alias, protectedPrivKey))

        elif (tag == 2):  # trusted certificate entry
            # Read the alias
            read_utf_string(fd)

            # Read the (entry creation) date
            buf = fd.read(8)

            # Read the trusted certificate
            if xVersion == 2:
                # read the certificate type
                read_utf_string(fd)

            buf = fd.read(4)
            certsize = struct.unpack("> I", buf)[0]
            certdata = fd.read(certsize)
            assert(len(certdata) == certsize)
        else:
            sys.stderr.write("%s: Unrecognized keystore entry\n" % filename)
            fd.close()
            return

    # how much data have we processed
    pos = fd.tell()

    # read hash
    md = fd.read(20)
    assert(len(md) == 20)

    # Output for container password (target 0)
    if keys:
        # Use the first key for the container password hash
        alias, protectedPrivKey = keys[0]
        keysize = len(protectedPrivKey)
        keydata = hexlify(protectedPrivKey).decode('ascii')
        nkeys = 1
    else:
        keysize = 0
        keydata = ""
        nkeys = 0

    sys.stdout.write("%s:$keystore$0$%d$%s" % (os.path.basename(filename), pos,
                                               hexlify(data[0:pos]).decode('ascii')))

    sys.stdout.write("$%s" % hexlify(md).decode('ascii'))
    sys.stdout.write("$%d$%d$%s" % (nkeys, keysize, keydata))
    sys.stdout.write(":::::%s\n" % filename)

    # Output for each key password (target 1)
    for alias, protectedPrivKey in keys:
        try:
            alias_str = alias.decode('utf-8', 'replace')
        except:
            alias_str = str(alias)

        # JtR label shouldn't contain ':'
        alias_str = alias_str.replace(':', '_')

        keysize = len(protectedPrivKey)
        keydata = hexlify(protectedPrivKey).decode('ascii')

        sys.stdout.write("%s-%s:$keystore$1$%d$%s" % (os.path.basename(filename), alias_str, pos,
                                                      hexlify(data[0:pos]).decode('ascii')))
        sys.stdout.write("$%s" % hexlify(md).decode('ascii'))
        sys.stdout.write("$1$%d$%s" % (keysize, keydata))
        sys.stdout.write(":::::%s\n" % filename)

    fd.close()


if __name__ == "__main__":
    if len(sys.argv) < 2:
        sys.stderr.write("Usage: %s <.keystore / .jks file(s)>\n" % sys.argv[0])
        sys.exit(1)

    for i in range(1, len(sys.argv)):
        process_file(sys.argv[i])
